README.md
# call
<div align="center">
    <a href="https://space.bilibili.com/264346349">
        <img src="https://img.shields.io/badge/bilibili-魔咔啦咔-blueviolet" alt="😭" />
    </a>
    <a href="https://github.com/ggdream/call">
        <img src="https://img.shields.io/badge/GitHub-魔咔啦咔-ff69b4" alt="😭" />
    </a>
</div>
<div align="center">
    <img src="https://img.shields.io/badge/Platforms-Windows,Linux,MacOS-009688" alt="😭" />
    <img src="https://img.shields.io/badge/Mode-Debug,Release-3949ab" alt="😭" />
</div>

You can use the package to open the dylib defined in the `assets` field of `pubsepc.yaml`.
<br/>
<br/>
<br/>


## 1.Prepare your dynamic link library
- Write your C-code
```c
// file: ${PROJECT_ROOT}/assets/main.c

int add(int a, int b) {
  return a + b;
}
```

- Compile it to a dylib
```sh
gcc -shared -fPIC -o libadd.so main.c           # Linux
gcc -shared -fPIC -o libadd.dll main.c          # Windows
clang -shared -fPIC -o libadd.dylib main.c      # MacOS

# file: ${PROJECT_ROOT}/assets/libadd.dll
```

## 2. Declare the assets path
You should declare path of the dylib in the `pubspec.yaml` file as images.
~~~yaml
flutter:
  assets:
    - assets/libadd.dll # Fill it in according to your storage location
~~~

## 3. Write flutter core code to call native function.
~~~dart
import 'package:flutter/material.dart';

import 'dart:ffi' as ffi;
import 'package:call/call.dart';

typedef FuncNative = ffi.Int32 Function(ffi.Int32, ffi.Int32);
typedef FuncDart = int Function(int, int);



void main() => runApp(App());

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<App> {
  @override
  Widget build(BuildContext context) {
    var dll = getDyLibModule('assets/libadd.dll');     // use it as images.
    var add = dll.lookupFunction<FuncNative, FuncDart>('add');

    return Text(
             add(999, 54639).toString(),
             textDirection: TextDirection.ltr
          );
  }
}
~~~

## 4. Run the flutter application

Finally, You can see the number `55638` in the top left corner of the application.


CHANGELOG.md
## [0.3.0] - Null Safety.
- Rewrite the doc on pub.dev

## [0.2.3] - Repair the doc.
- Repair the url path of GitHub README.md file.

## [0.2.2] - Repair the travis-ci file.

## [0.2.1] - Repair the doc.
- Add travis-ci supported.

## [0.2.0] - Change internal API.
- From `interface` to `abstract class`.
- Add [getDyLibModule] code docs to read API easily.

## [0.1.0] - Change internal API.
- Support MacOS Release Mode.

## [0.0.3] - Support Linux.
Linux test.

## [0.0.2] - README.
Add docs content.

## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.


test/call_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:call/call.dart';

void main() {
  // test('adds one to input values', () {
  //   getDyLibModule('assets/libtest.dll');
  // });
  print('call, hello');
}


lib/core.dart
import 'dart:io' show Directory;
import 'package:path/path.dart' show join;

import 'env.dart';

abstract class TargetPath {
  String getDebugPath();
  String getReleasePath() =>
      join(Directory.current.path, 'data', 'flutter_assets');
  String getCurrentPath() => isProduct() ? getReleasePath() : getDebugPath();
}

class Windows extends TargetPath {
  @override
  String getDebugPath() => join(Directory.current.path,
      'build\\windows\\runner\\Debug', 'data', 'flutter_assets');
}

class Linux extends TargetPath {
  @override
  String getDebugPath() => join(Directory.current.path,
      'build/linux/debug/bundle', 'data', 'flutter_assets');
}

class Macos extends TargetPath {
  @Deprecated(
      "I'm sorry that i doesn't have a Mac, so doesn't know the path relationship in debug mode")
  @override
  String getDebugPath() {
    throw UnimplementedError();
  }
}


lib/meta.dart
import 'dart:io' show Platform;
import 'package:path/path.dart' show dirname, join;

import 'env.dart';



abstract class Meta {
  String getDebugPath();
  String getReleasePath();
  String getCurrentPath();
}

class Windows implements Meta {
  static const String assets = 'data\\flutter_assets';

  @override
  String getDebugPath() => join(
      dirname(Platform.script.toFilePath(windows: true)),
      'build\\windows\\runner\\Debug',
      assets);

  @override
  String getReleasePath() =>
      join(dirname(Platform.script.toFilePath(windows: true)), assets);

  @override 
  String getCurrentPath() => isProduct() ? getReleasePath() : getDebugPath();
}

class Linux implements Meta {
  static const String assets = 'data/flutter_assets';

  @override
  String getDebugPath() => join(
      dirname(Platform.script.toFilePath(windows: false)),
      'build/linux/debug/bundle',
      assets);

  @override
  String getReleasePath() =>
      join(dirname(Platform.script.toFilePath(windows: false)), assets);

  @override
  String getCurrentPath() => isProduct() ? getReleasePath() : getDebugPath();
}

class Macos implements Meta {
  @override
  String getDebugPath() {
    // TODO: implement getDebugPath
    throw UnimplementedError();
  }

  @override
  String getReleasePath() {
    // TODO: implement getReleasePath
    throw UnimplementedError();
  }

  @override
  String getCurrentPath() {
    // TODO: implement getCurrentPath
    throw UnimplementedError();
  }
}


lib/ffi.dart
import 'dart:ffi' as ffi;
import 'dart:io' show Platform;
import 'package:path/path.dart' show join;

import 'core.dart';

/// You can use the [getDyLibModule] function to open the dylib defined in the `assets` field of `pubsepc.yaml`.
/// 
/// For example:
/// clang code
/// ```
/// int add(int a, int b) {
///   return a + b;
/// }
///
/// // gcc -shared -fPIC -o libadd.dll main.c     (Windows)
/// // gcc -shared -fPIC -o libadd.so main.c      (Linux)
/// // clang -shared -fPIC -o libadd.dylib main.c (MacOS)
/// 
/// // Finally: {PROJECT_ROOT}/assets/libadd.dll (must be declared in `pubsepc.yaml` file)
/// ```
/// 
/// 
///Dart code
/// ```
/// import 'dart:ffi' as ffi;
/// import 'package:call/call.dart';
///
/// typedef FuncNative = ffi.Int32 Function(ffi.Int32, ffi.Int32);
/// typedef FuncDart = int Function(int, int);
///
/// var dll = getDyLibModule('assets/libadd.dll');
/// var add = dll.lookupFunction<FuncNative, FuncDart>('add');
///
/// print(add(999, 54639));     // Output: 55638
/// ```
ffi.DynamicLibrary getDyLibModule(String module) {
  String path = '';
  if (Platform.isWindows) {
    path = Windows().getCurrentPath();
  } else if (Platform.isLinux) {
    path = Linux().getCurrentPath();
  } else if (Platform.isMacOS) {
    path = Macos().getCurrentPath();
  } else {
    throw 'The version lib doesn\'t support the platform';
  }

  return ffi.DynamicLibrary.open(join(path, module));
}
